nv2a: Fix screen clip coordinate calculation when reusing surfaces with different clip bounds

lostromb opened the pull request:
Factor in the difference between surface and clip size when calculating clip coords.

This should be considered a hack. I have no justification for this change, and no hardware tests to support it. But it does fix splitscreen rendering in Halo 2 and Phantom Dust without regressing any other game I've tested.

Master branch:
![image](https://github.com/user-attachments/assets/17045f22-6c0d-427f-989c-5842a667c221)

This PR:
![image](https://github.com/user-attachments/assets/dbe18824-75d3-4698-9ace-17d4e2bc4bda)

!fixes #1986

MasonT8198 commented:
Can you run this against abaires test cases to see if this breaks anything? I feel like it probably will

lostromb commented:
The only hardware I have to test with is an 8-year-old Quadro P400. I get a segfault about 10 seconds into the suite. Also the first time I ever saw the Xemu codebase was this morning.

mborgerson commented:
Thanks for the debugging effort and PR!

If you want to help build some surface tests (or check for existing ones, maybe some exist since last I checked) it will help build lasting support for the PR and reduce risk of future regression.

I'm aware some tests cause xemu to crash, we will fix them in due time. But until then, you can disable those tests that are crashing on xemu master branch.

abaire commented:
> I'm aware some tests cause xemu to crash, we will fix them in due time. But until then, you can disable those tests that are crashing on xemu master branch.

Or you can try out https://github.com/abaire/xemu-dev_pgraph_test_results which will proceed past the crashes and generate a page with all of the diffs caused by this PR. Instructions are in the README. 

So far I think I'm the only one who has used it, so there may be rough edges to work out; feel free to let me know if you run into trouble.

lostromb commented:
Thank you for that. First off, this definitely still has problems as 2-player mode is now broken in the opposite way as before.
![image](https://github.com/user-attachments/assets/54fefb68-a8ee-4f8e-92e1-010a1faa3b72)

I'll try and dig into that next and see what's the difference between the draw calls in this configuration.

I was able to get Win64 crossbuild working so I can test on a better graphics card. I can run individual pgraph tests now, but I don't actually see any coverage for scissor testing / clip rectangles in particular? There's cases for stencil buffers and viewports, but I don't think either of those apply directly to the changes I'm making.

abaire commented:
> ... I don't actually see any coverage for scissor testing / clip rectangles in particular? There's cases for stencil buffers and viewports, but I don't think either of those apply directly to the changes I'm making.

Yeah, the test coverage is nowhere exhaustive. Most of these tests are just related to some specific bug/feature I (or a contributor) was working on at the time. As Matt suggested above, it'd be excellent if you could add tests that reproduce the issue you're fixing along with this PR. You can ping me here or on Discord if you run into issues; I've recently started adding some documentation for the test infra but it's still sparse.

lostromb commented:
This PR also fixes Phantom Dust splitscreen, which might be notable as that's the only other splitscreen game I've found that uses clip coordinates that are not just (screen size divided by 2).

Note how each player's view spills over to the left a bit on master, this is corrected in the PR.
![image](https://github.com/user-attachments/assets/95056955-11e9-486a-80dd-f663dde438e2)

lostromb commented:
Other games are unremarkable.

4x4 EVO
![image](https://github.com/user-attachments/assets/948434b8-7215-4dec-9a6e-b9c3711d9eeb)

Aggressive Inline
![image](https://github.com/user-attachments/assets/f277cb37-38eb-4bdc-96a5-032d41409dee)

Halo CE
![image](https://github.com/user-attachments/assets/686fc0db-3fed-4fd6-bd1b-a0768a9ea41e)

Halo 2
![image](https://github.com/user-attachments/assets/a0bedaee-d50a-4008-883c-0b13665d2eb9)

MX Superfly
![image](https://github.com/user-attachments/assets/908b8cf3-60a2-4469-8620-bedc6e40a3c9)

Project Gotham
![image](https://github.com/user-attachments/assets/c9303755-81cf-4bbd-80f5-c0531e1fd4f5)

mborgerson commented:
Do you plan to create test cases for this change?

lostromb commented:
I would like to, of course. Haven't been able to build the pgraph test project yet. The cases themselves should be very straightforward - basically copies of the viewport tests except setting clip registers instead of viewport bounds. The game tests yesterday confirmed that this might actually be a more legitimate fix than the hack I first though it was. Still just kinda poking around with stuff and learning the tools

mborgerson commented on hw/xbox/nv2a/pgraph/gl/draw.c:
```diff
@@ -332,8 +332,10 @@ void pgraph_gl_draw_begin(NV2AState *d)
     /* FIXME: Consider moving to PSH w/ window clip */
-    unsigned int xmin = pg->surface_shape.clip_x - pg->surface_binding_dim.clip_x,
-                 ymin = pg->surface_shape.clip_y - pg->surface_binding_dim.clip_y;
+    unsigned int xmin = (pg->surface_shape.clip_x - pg->surface_binding_dim.clip_x) +
```
I think this should simplify to `xmin = pg->surface_shape.clip_x;`, since clip offset is included in surface dims now. Similarly for ymin.

https://github.com/xemu-project/xemu/blob/fac9ae83c7f673f5991ab03217226f0b97b3578b/hw/xbox/nv2a/pgraph/gl/surface.c#L1046-L1047

lostromb replied:
I should have listened to this in the first place because it's exactly correct

mborgerson commented:
> I would like to, of course.

Great, thanks! I'll await the surface tests cases you are creating before merging the change then. In the meantime I'll move this PR into draft state.

lostromb commented:
Turns out there's already a little bit of coverage - the final two test cases under Surface Clip. I didn't notice the few red pixels on the side in these specific cases.

Master
![image](https://github.com/user-attachments/assets/a966a7e3-472c-46d0-a5c8-3bdb3de399db)

PR
![image](https://github.com/user-attachments/assets/749d09c4-4382-44cc-a3c5-c00568a36a47)

MasonT8198 commented:
> Turns out there's already a little bit of coverage - the final two test cases under Surface Clip. I didn't notice the few red pixels on the side in these specific cases.
> 
> Master ![image](https://private-user-images.githubusercontent.com/918053/422588819-a966a7e3-472c-46d0-a5c8-3bdb3de399db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE5MTQyMzIsIm5iZiI6MTc0MTkxMzkzMiwicGF0aCI6Ii85MTgwNTMvNDIyNTg4ODE5LWE5NjZhN2UzLTQ3MmMtNDZkMC1hNWM4LTNiZGIzZGUzOTlkYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxNFQwMDU4NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jZjZmMDhjM2M0ZTY5NjczNjBhMDM2MDQ4ODc0NjAwYThkMTE2NzZmOWM5ZmJkNGM4OWQyMTQyYjU2MTRiMjg1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.FUZZvU6ls91x60vbxnLL657i7vcqaPdPpSRZ0mREzNU)
> 
> PR ![image](https://private-user-images.githubusercontent.com/918053/422588861-749d09c4-4382-44cc-a3c5-c00568a36a47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE5MTQyMzIsIm5iZiI6MTc0MTkxMzkzMiwicGF0aCI6Ii85MTgwNTMvNDIyNTg4ODYxLTc0OWQwOWM0LTQzODItNDRjYy1hM2M1LWMwMDU2OGEzNmE0Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxNFQwMDU4NTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZjVlODMzNTQ1NjY5NzU1MWQwMTVmMWQ5MzViZWQzMmRiMDUzOWM0N2UyY2E4NDU4ZjljNjQ3YWUyMjRhZjIzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cenKMFbUFgKRpm3RzpiG2PUCesNRl34QikHc-rPz58o)

Abaire has a repository with golden results, compare to that just to see what the difference is

lostromb commented:
> > I'm aware some tests cause xemu to crash, we will fix them in due time. But until then, you can disable those tests that are crashing on xemu master branch.
> 
> Or you can try out https://github.com/abaire/xemu-dev_pgraph_test_results which will proceed past the crashes and generate a page with all of the diffs caused by this PR. Instructions are in the README.
> 
> So far I think I'm the only one who has used it, so there may be rough edges to work out; feel free to let me know if you run into trouble.

I don't know enough python to figure this out, unfortunately...
```
./.venv/bin/pip -r requirements.txt
ERROR: Could not find a version that satisfies the requirement nxdk-pgraph-test-runner>=0.1.12 (from versions: none)
ERROR: No matching distribution found for nxdk-pgraph-test-runner>=0.1.12
```

abaire commented:
What version of Python are you using (and what platform are you on)?


The test being inconclusive just means a better test probably to be written, I don't recall what I was working on when I wrote those tests, but I probably wasn't doing a very deep dive on the behavior.

According to https://abaire.github.io/xemu-nxdk_pgraph_tests_results/compare/xemu-0.8.34-master-a143f66ce48ef37ccb15bbdd4daed3c6df2eaee9/Darwin_arm64/gl_Apple_Apple_M3_Max%3Agslv_4.10/Surface_clip.html there are some existing differences on master (Darwin/OpenGL) vs Xbox so maybe the test is good enough but could be made more obvious?

lostromb commented:
> According to https://abaire.github.io/xemu-nxdk_pgraph_tests_results/compare/xemu-0.8.34-master-a143f66ce48ef37ccb15bbdd4daed3c6df2eaee9/Darwin_arm64/gl_Apple_Apple_M3_Max%3Agslv_4.10/Surface_clip.html there are some existing differences on master (Darwin/OpenGL) vs Xbox so maybe the test is good enough but could be made more obvious?

Yeah, I think if you just added one more case with like `x=160,y=80,w=300,h=200` then the difference would become super obvious.
The Halo 2 behavior should be `x=0,y=240,w=640,h=240` so you could add that too

mborgerson commented:
> Yeah, I think if you just added one more case with like `x=160,y=80,w=300,h=200` then the difference would become super obvious.
> The Halo 2 behavior should be `x=0,y=240,w=640,h=240` (and for 4-way, `x=320,y=240,w=320,h=240`) so you could add those too

@lostromb You can add the necessary test cases and send a PR with them to https://github.com/abaire/nxdk_pgraph_tests

abaire commented:
> @lostromb You can add the necessary test cases and send a PR with them to https://github.com/abaire/nxdk_pgraph_tests

Normally I would've said the same thing, but I need to add the new doc comments to those tests anyway so I've already incorporated the additional regions (and added tests for 32bpp surfaces). I'm waiting on a HW regression, but you can hopefully test the ISO from  https://github.com/abaire/nxdk_pgraph_tests/pull/200 (or wait an hour or so until I merge the change)

MasonT8198 commented:
Abaire do you mind grabbing the golden on your devkit and posting it? My
xboxen are in storage and I have yet to go grab them (as in I forgot where
I put the key to my storage building on my back lawn)

On Fri, Mar 14, 2025 at 12:14 AM Erik Abair ***@***.***>
wrote:

> @lostromb <https://github.com/lostromb> You can add the necessary test
> cases and send a PR with them to
> https://github.com/abaire/nxdk_pgraph_tests
>
> Normally I would've said the same thing, but I need to add the new doc
> comments to those tests anyway so I've already incorporated the additional
> regions (and added tests for 32bpp surfaces). I'm waiting on a HW
> regression, but you can hopefully test the ISO from
> abaire/nxdk_pgraph_tests#200
> <https://github.com/abaire/nxdk_pgraph_tests/pull/200> (or wait an hour
> or so until I merge the change)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemu-project/xemu/pull/1998#issuecomment-2723393691>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARHXUZO653WPA7CXZAJKZKT2UJJTBAVCNFSM6AAAAABY2U3LACVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRTGM4TGNRZGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: abaire]*abaire* left a comment (xemu-project/xemu#1998)
> <https://github.com/xemu-project/xemu/pull/1998#issuecomment-2723393691>
>
> @lostromb <https://github.com/lostromb> You can add the necessary test
> cases and send a PR with them to
> https://github.com/abaire/nxdk_pgraph_tests
>
> Normally I would've said the same thing, but I need to add the new doc
> comments to those tests anyway so I've already incorporated the additional
> regions (and added tests for 32bpp surfaces). I'm waiting on a HW
> regression, but you can hopefully test the ISO from
> abaire/nxdk_pgraph_tests#200
> <https://github.com/abaire/nxdk_pgraph_tests/pull/200> (or wait an hour
> or so until I merge the change)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemu-project/xemu/pull/1998#issuecomment-2723393691>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARHXUZO653WPA7CXZAJKZKT2UJJTBAVCNFSM6AAAAABY2U3LACVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRTGM4TGNRZGE>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

lostromb commented:
> > @lostromb You can add the necessary test cases and send a PR with them to https://github.com/abaire/nxdk_pgraph_tests
> 
> Normally I would've said the same thing, but I need to add the new doc comments to those tests anyway so I've already incorporated the additional regions (and added tests for 32bpp surfaces). I'm waiting on a HW regression, but you can hopefully test the ISO from [abaire/nxdk_pgraph_tests#200](https://github.com/abaire/nxdk_pgraph_tests/pull/200) (or wait an hour or so until I merge the change)

Thank you much for helping out 👍. The difference in test outputs should be very obvious now with the new cases. I eagerly await the hardware parity results.

Master branch
![image](https://github.com/user-attachments/assets/ede081c1-2663-42e2-952a-68a6d4cbab27)

PR
![image](https://github.com/user-attachments/assets/28640e0a-ec34-41a7-882c-ce97bc15a2cf)

In the meantime I tested every other splitscreen game I could find and couldn't tell any differences in any of them.

abaire commented:
The results are already up https://abaire.github.io/nxdk_pgraph_tests_golden_results/results/Surface_clip/index.html and the PR output does not look correct to me. 

It'd be good to figure out why the dev test script is failing for you so you can run a full regression. What Python version are you using and what OS?

lostromb commented:
Okay, the new pgraph tests are great because they confirm my hunch that this PR's "fix" is still kind of a hack that's actually just canceling out a miscalculation which happened earlier. The fact that the clip boundaries change based on the surface format was the main tip-off. With format RGB565 the clip gets messed up, but with RGBA8888 it works as expected.

The root cause, I believe, is in surface.c when we're trying to look around to see if we can reuse an existing surface rather than recreating it. When we find a match, inherit the properties from the found match, including clip bounds, and reuse that surface instead of recreating it. The problem arises when we've got our own set of clip bounds that don't match with the clip bounds set in the found surface.

I assume based on #2005 that the ideal scenario would be to allow surface reuse even for when clip bounds don't match rather than always recreate the surface, so I think it's just a matter of correctly reconciling those bounds.

> It'd be good to figure out why the dev test script is failing for you so you can run a full regression. What Python version are you using and what OS?

Let's discuss at https://github.com/abaire/xemu-dev_pgraph_test_results/issues/9

lostromb commented:
I think the fixes are in the correct place now, games look fine, the pgraph clip tests all look good to me.
Batch test results PR is here https://github.com/abaire/xemu-dev_pgraph_test_results/pull/10

abaire commented:
> I think the fixes are in the correct place now, games look fine, running the pgraph clip tests manually all look fine to me. Batch test results PR is here [abaire/xemu-dev_pgraph_test_results#10](https://github.com/abaire/xemu-dev_pgraph_test_results/pull/10)

Unfortunately the GH action has some issues running on PRs not authored by the repo autho so it didn't generate the comparison page. I think if you create the PR against your own repo's main branch it'll run there and you can provide a link to the results page (it'll look something like https://abaire.github.io/xemu-dev_pgraph_test_results/feat_support_alpha_from_specular/index.html )

I updated the instructions on the dev repo to hopefully clarify for future users; sorry for the inconvenience!

lostromb commented:
> Unfortunately the GH action has some issues running on PRs not authored by the repo owner so it didn't generate the comparison page. I think if you create the PR against your own repo's main branch it'll run there and you can provide a link to the results page (it'll look something like https://abaire.github.io/xemu-dev_pgraph_test_results/feat_support_alpha_from_specular/index.html )

???? https://github.com/lostromb/xemu-dev_pgraph_test_results/pull/1

In other news, this change feels incredible
https://www.youtube.com/watch?v=NrDpNfDCsuM

lostromb commented:
Okay, so you need to
- Manually enable workflows in the forked repository
- Manually setup Github Pages to use workflows as you described
- Manually create the wiki repo by creating a dummy page
- Must have a branch called "master", so just "main" won't work (Possibly?)

With all that, the github pages still don't deploy, but I got a wiki page that is mostly useless https://github.com/lostromb/xemu-dev_pgraph_test_results/wiki/screen_clip_fix
The comparison output still seems to hardcode the repository path as "abaire/whateverreponame" so it's trying to pull the results from the PR and never finding them.

For now it seems like it would be easier to pull [my results branch](https://github.com/lostromb/xemu-dev_pgraph_test_results/tree/screen_clip_fix) into your repository and just run the actions from there.

abaire commented:
Thanks for testing!

I assume it's not literally "abaire/whateverreponame" and depending on what "whateverreponame" actually is, it may be correct. The golden images are all accessed in other repos under my account.

I removed wiki generation entirely as it's not very useful anyway. I also updated the instructions to try to capture all the first time setup steps and added some debug logging to try to track down why the site didn't generate.

You're welcome to grab the updates and retry (probably just need the [`build.yml`](https://github.com/abaire/xemu-dev_pgraph_test_results/blob/main/.github/workflows/build.yml). I'll see if I can do some testing with another account to work out the remaining kinks.

lostromb commented:
Okay, I also just don't have the hardware configuration to run an adequate test right now, unless I could get the test harness working on Windows. Or I could reformat another spare storage box to Linux and throw a GTX 1550 in there.... This might take a while.

abaire commented:
> Okay, I also just don't have the hardware configuration to run an adequate test for this PR right now, unless I could get the test harness working on Windows. Or I could reformat another spare storage box to Linux and throw a GTX 1550 in there.... This might take a while.

In theory it should run on Windows (the underlying packages do, since that's what I use to gen the xemu golden results). I'm traveling for a bit and won't have access to a Windows machine but happy to provide suggestions if you run into problems there. In my case I have Python3.11 installed and run the test suite from cygwin.

That said, you've already generated the diffs, so all of the actual processing (the comparisons vs goldens, site generation) should be happening via the GitHub actions, so I think all you need to get the diffs working is git. 
If you're still working on the xemu PR you should be able to use the pre-built pgraph test ISO without having to spend time getting that building. You can also create a JSON config file to prevent autorun and skip all the tests besides the "Surface clip" and throw it on the E: to speed up your debug cycle (instructions in the repo, you may be the first person besides me to try them so let me know if you try and hit trouble so I can improve them).

lostromb commented:
Okay, well I fixed up the wiki page well enough that you can at least view the diffs in an inconvenient way. [Here are the surface clip results.](https://github.com/lostromb/xemu-dev_pgraph_test_results/wiki/screen_clip_fix#surface_clip). But I would assume all of my other differences are a result of doing the test on an old Quadro that can't even run the Vulkan renderer.

MasonT8198 commented:
Weird that Swath Width is invisible via PR

On Sat, Mar 15, 2025 at 4:37 PM Logan Stromberg ***@***.***>
wrote:

> Okay, well I fixed up the wiki page well enough that you can at least view
> the diffs in an inconvenient way. Here are the surface clip results.
> <https://github.com/lostromb/xemu-dev_pgraph_test_results/wiki/screen_clip_fix#surface_clip>.
> But I would assume all of my other differences are a result of doing the
> test on an old Quadro that can't even run the Vulkan renderer.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemu-project/xemu/pull/1998#issuecomment-2726989833>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARHXUZM6WDKZHE6NFWHLJDT2USFPJAVCNFSM6AAAAABY2U3LACVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWHE4DSOBTGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
> [image: lostromb]*lostromb* left a comment (xemu-project/xemu#1998)
> <https://github.com/xemu-project/xemu/pull/1998#issuecomment-2726989833>
>
> Okay, well I fixed up the wiki page well enough that you can at least view
> the diffs in an inconvenient way. Here are the surface clip results.
> <https://github.com/lostromb/xemu-dev_pgraph_test_results/wiki/screen_clip_fix#surface_clip>.
> But I would assume all of my other differences are a result of doing the
> test on an old Quadro that can't even run the Vulkan renderer.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/xemu-project/xemu/pull/1998#issuecomment-2726989833>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ARHXUZM6WDKZHE6NFWHLJDT2USFPJAVCNFSM6AAAAABY2U3LACVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMRWHE4DSOBTGM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

abaire commented:
> **In all surface clip tests the PR matches hardware**, except for what looks like a separate issue where the clipping interacts with the test name overlay itself (maybe xemu doesn't apply clipping rects to blittable areas or something)

Yeah, confirmed that this is an issue at master. I'm adding a test specifically for this case and fixing the text rendering in the rest of the surface clip tests.

> Weird that Swath Width is invisible via PR

Looks like it's not just Swath Width. It'd be good to check master on the same hardware to verify that this is an issue unrelated to the PR.

lostromb commented:
[Got a proper test result using a newer Nvidia card on Linux](https://lostromb.github.io/pgraph_test_results_2/site/screen_clip_test_gtx1660/index.html), but it doesn't appear any better than the last one.

abaire commented:
Just to clarify, are you testing across machines with an unchanged build at master, or just with the changes in this branch?

lostromb commented:
> Just to clarify, are you testing across machines with an unchanged build at master, or just with the changes in this branch?

I'm moving this branch's dev build to different machines and manually looking at individual test results since I couldn't get the automated script to run in Cygwin.

I added an ugly-looking patch to the code to avoid negative clips, these are the diffs now. The list of changes is much smaller.
https://lostromb.github.io/pgraph_test_results_2/site/screen_clip_test_7dbbef1/

lostromb commented:
Okay the correct fix is actually a lot simpler than I was making it out to be. 4 lines changed. @mborgerson this should be ready for review.

[Pgraph test results](https://lostromb.github.io/pgraph_test_results_2/site/screen_clip_test_4df526b/)

mborgerson commented:
Thanks!
